proposals/0351-regex-builder.md

```diff
@@ -428,16 +441,204 @@ extension RegexComponentBuilder {
 
 `buildOptional` and `buildEither` are intentionally not supported due to ergonomic issues and fundamental semantic differences between regex conditionals and result builder conditionals. Please refer to the [alternatives considered](#support-buildoptional-and-buildeither) section for detailed rationale.
 
+</details>
+
+### Capture and reference
+
+Capture is a common regex feature that saves a portion of the input upon match. In regex builder, `Capture` and `TryCapture` are regex components that produce a new regex by inserting the captured pattern's whole match (`.0`) to the `.1` position of `Output`. When a transform closure is provided, the whole match (`.0`) of the captured content will be transformed to using the closure.
+
+```swift
+public struct Capture<Output>: RegexComponent { ... }
+public struct TryCapture<Output>: RegexComponent { ... }
+```
+
+To do a simple capture, you provide `Capture` with a regex component or a regex component builder closure.
+
+```swift
+// Equivalent: '(CREDIT|DEBIT)'
+Capture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} // `.RegexOutput == (Substring, Substring)`
+```
+
+To transform the captured substring, specify a `transform:` closure.
+
+```
+// This example is similar to the one above, however in this example we
+// transform the result of the capture into:
+// "Transaction Kind: CREDIT" or "Transaction Kind: DEBIT"
+Capture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  "Transaction Kind: \($0)"
+} // `.RegexOutput == (Substring, String)`
+```
+
+The transform closure can throw. When a transform closure throws during matching, the matching will abort and the error will be propagated directly to the top-level matching API that's being called, e.g. `Regex.wholeMatch(in:)` and `Regex.prefixMatch(in:)`.
+
+An alternative version of capture is called `TryCapture`, which works in cases where you want to transform the capture, but the transformation may return nil. When a nil is returned, the regex engine backtracks and tries an alternative. `TryCapture` is most useful when you transform a captured substring by calling a failable initializer.
+
+```swift
+enum TransactionKind: String {
+  case credit = "CREDIT"
+  case debit = "DEBIT"
+}
+
+TryCapture {
+  ChoiceOf {
+    "CREDIT"
+    "DEBIT"
+  }
+} transform: {
+  // This initializer may return nil which is why we used TryCapture.
+  TransactionKind(rawValue: String($0))
+}
+```
+
```

Comment:
Maybe make reference treatment a sub-section here? Or should we finish non-reference captures here and make this the start of the next section?

Reply:
I made reference a new section after `Capture`.